src/coreclr/jit/lower.cpp

```diff
@@ -1270,20 +1270,32 @@ bool Lowering::TryLowerSwitchToBitTest(
         comp->fgAddRefPred(bbCase1, bbSwitch);
     }
 
+    var_types bitTableType = (bitCount <= (genTypeSize(TYP_INT) * 8)) ? TYP_INT : TYP_LONG;
+    GenTree*  bitTableIcon = comp->gtNewIconNode(bitTable, bitTableType);
+
+#ifdef TARGET_XARCH
     //
     // Append BT(bitTable, switchValue) and JCC(condition) to the switch block.
     //
-
-    var_types bitTableType = (bitCount <= (genTypeSize(TYP_INT) * 8)) ? TYP_INT : TYP_LONG;
-    GenTree*  bitTableIcon = comp->gtNewIconNode(bitTable, bitTableType);
-    GenTree*  bitTest      = comp->gtNewOperNode(GT_BT, TYP_VOID, bitTableIcon, switchValue);
+    GenTree* bitTest = comp->gtNewOperNode(GT_BT, TYP_VOID, bitTableIcon, switchValue);
     bitTest->gtFlags |= GTF_SET_FLAGS;
     GenTreeCC* jcc = comp->gtNewCC(GT_JCC, TYP_VOID, bbSwitchCondition);
-
     LIR::AsRange(bbSwitch).InsertAfter(switchValue, bitTableIcon, bitTest, jcc);
+#else  // TARGET_XARCH
+    //
+    // Fallback to AND(RSZ(bitTable, switchValue), 1)
+    //
+    GenTree* tstCns  = comp->gtNewIconNode(bbSwitch->bbNext == bbCase0 ? 0 : 1, bitTableType);
+    GenTree* shift   = comp->gtNewOperNode(GT_RSZ, bitTableType, bitTableIcon, switchValue);
+    GenTree* bitTest = comp->gtNewOperNode(GT_TEST, TYP_VOID, shift, tstCns);
+    bitTest->gtFlags |= GTF_SET_FLAGS;
+    GenTreeCC* jcc = comp->gtNewCC(GT_JCC, TYP_VOID, GenCondition::EQ);
+    LIR::AsRange(bbSwitch).InsertAfter(switchValue, bitTableIcon, shift, tstCns, bitTest);
+    LIR::AsRange(bbSwitch).InsertAfter(bitTest, jcc);
```

Comment:
I would create `JTRUE` and continue lowering from the `JTRUE`, agnostically on all platforms. I believe it will produce more optimal codegen for ARM64 (`tbnz`) and the same codegen on x64.

Reply:
Also this logic (with `GT_TEST`) likely does not work on RISC-V and LA64. So it's best to just create normal IR (regular ANDs) and leave it up to regular lowering to optimize it.

Reply:
Good point, that's why I enabled the whole thing only for XARCH and ARM64 (arm32 doesn't support TEST), will try your suggestion